### PR TITLE
Merge feature/pbst-0002-adding-server-start-stop-command

### PR DESCRIPTION
**Pull request by j-stahl**
Added start and stop command for being able to switch VMs on and off via cli.

### Screenshot:
grafik

### Tests done:
- \[x] starting a VM via vagrant
- \[x] stopping a VM via vagrant
- \[x] giving a wrong argument (server name) to the command

### Squash message:
* Added feature start command
* Added feature stop command

### CODE REVIEW

File: commands/serverstartcommand.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    protected function configure()
+    {
+        parent::configure();
+        $this
+            ->setName("server:start")
+            ->setDescription("Starts one or more profitbricks server!")
+            ->addArgument("serverName", InputArgument::IS_ARRAY, "Names or IDs of server, which should be started.");
+    }
+
+    protected function interact(InputInterface $input, OutputInterface $output)
```
Comment (Shyru) [inline commands/serverstartcommand.php]: What is this used for? To check if the argument is passed into the input? You know that an argument can be defined as required and symfony command will moan if the argument is not present?

File: profitbricksapi/profitbricksapi.php [hunk 2]
```diff
@@ -25,6 +25,8 @@ class ProfitBricksApi
     const profitBricksDiskApi = "https://api.profitbricks.com/cloudapi/v3/datacenters/[data-center-id]/servers/[server-id]/volumes?depth=1";
     const profitBricksDeleteSnapshot = "https://api.profitbricks.com/cloudapi/v3/snapshots/[snapshot-id]";
     const profitBricksCreateSnapshot = "https://api.profitbricks.com/cloudapi/v3/datacenters/[data-center-id]/volumes/[volume-id]/create-snapshot";
+    const profitBricksServerStart = "https://api.profitbricks.com/cloudapi/v4/datacenters/{dataCenterId}/servers/{serverId}/start";
+    const profitBricksServerStop = "https://api.profitbricks.com/cloudapi/v4/datacenters/{dataCenterId}/servers/{serverId}/stop";
```
Comment (Shyru) [inline profitbricksapi/profitbricksapi.php:29]: Please use `[]` here also for placeholders like in the other urls, or change all to `{}`.

**Review by Shyru — CHANGES_REQUESTED** [review]
Looks generally good to me! :+1:
However overall there are quite some things that could/should be refactored.
You decide when you want to tackle these. (I ordered them in descending priority)
1. Use a trait or intermediary base-command that contains the code for finding virtual machines by name or id. There is a lot of duplicated code in the start/stop command that could be shared, when this is refactored into a trait/base-class.
2. Implement a private prepareUrl() method inside ProfitbricksAPI that does all the str-replace stuff on the urls so that the code can be reused more easily. Also you can use a single str_replace call and pass arrays with things to replace as first and second parameter.
3. Use [Httpful](http://web.archive.org/web/20180829114952/http://phphttpclient.com/)* as library to do the api-calls to Profitbricks. Would make the code much nicer to read and would avoid lots of duplicated curl_setopt()-calls. (See Templates for calls in the Httpful examples)

*Somehow the current page is looking really bad and contains all sorts of advertisment, looks like it has been hacked. The link above shows the page how it was before and looks much nicer.

File: commands/serverstartcommand.php (hunk 1 above)
Comment (j-stahl) [reply commands/serverstartcommand.php]: I knew about that. But I did not recognize, that there is the possibility to bitwise add the flags for having both options at once (ARRAY and REQUIRED).

File: profitbricksapi/profitbricksapi.php (hunk 2 above)
Comment (j-stahl) [reply profitbricksapi/profitbricksapi.php:29]: This results from the origin API documentation. I like the curly brakets more.

File: commands/servercommandbase.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ *      - provide a valid name, check/parse of new input arguments, arbitrary setup code before running the command
+ *  - Implement method execute() with what gets done if the command is triggered
+ */
+class ServerCommandBase extends CommandBase
+{
+    const on = 1;
+    const off = 2;
+
+    /**
+     * @param integer $_serverState
```
Comment (Shyru) [inline commands/servercommandbase.php]: Please document this parameter. (That it must be one of the consts on/off)

File: commands/servercommandbase.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+     * @param InputInterface $input
+     * @param OutputInterface $output
```
Comment (Shyru) [inline commands/servercommandbase.php]: Pelase also document these parameters and change them to $_input and $_output.

File: commands/servercommandbase.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Base-Class for all IONOS-related server commands.
+ *
+ * We inherit from our own base class instead direct use of symfony command because there are some lines of code
+ * which has to be executed always. All server commmands should use this class for a base class.
+ *  - Implement method configure() with command-specific additions as needed
+ *      - provide a valid name, check/parse of new input arguments, arbitrary setup code before running the command
+ *  - Implement method execute() with what gets done if the command is triggered
+ */
```
Comment (Shyru) [inline commands/servercommandbase.php:28]: Please document that the inheriting commands need an argument serverName that means the server that the command targets.

**Review by Shyru — CHANGES_REQUESTED** [review]
Good work in reusing the code! Please improve the documentation a bit, and then we are good to go!

**Review by Shyru — APPROVED** [review]
Good work!

**Pull request merged**